packages/explorer/src/observer/decorator.ts

```diff
@@ -19,52 +21,100 @@ export function observer({ explorerUrl = "http://localhost:13690", waitForTransa
   account extends Account | undefined = Account | undefined,
 >(
   client: Client<transport, chain, account>,
-) => Pick<WalletActions<chain, account>, "writeContract"> {
+) => Pick<WalletActions<chain, account>, "writeContract"> & Pick<BundlerActions, "sendUserOperation"> {
   const emit = createBridge({ url: `${explorerUrl}/internal/observer-relay` });
 
-  return (client) => ({
-    async writeContract(args) {
-      const writeId = `${client.uid}-${++writeCounter}`;
-      const write = getAction(client, writeContract, "writeContract")(args);
-
-      // `writeContract` above will throw if this isn't present
-      const functionAbiItem = getAbiItem({
-        abi: args.abi,
-        name: args.functionName,
-        args: args.args,
-      } as never)!;
-
-      emit("write", {
-        writeId,
-        address: args.address,
-        from: client.account!.address,
-        functionSignature: formatAbiItem(functionAbiItem),
-        args: (args.args ?? []) as never,
-        value: args.value,
-      });
-      Promise.allSettled([write]).then(([result]) => {
-        emit("write:result", { ...result, writeId });
-      });
-
-      write.then((hash) => {
-        const receipt = getAction(client, waitForTransactionReceipt, "waitForTransactionReceipt")({ hash });
-        emit("waitForTransactionReceipt", { writeId, hash });
-        Promise.allSettled([receipt]).then(([result]) => {
-          emit("waitForTransactionReceipt:result", { ...result, writeId });
+  return (client) => {
+    return {
+      async sendUserOperation(args) {
+        const writeId = `${client.uid}-${++writeCounter}`;
+        const write = getAction(client, sendUserOperation, "sendUserOperation")(args);
```

Comment:
I was looking at this and finding the many instances of an early return with the same value (`return write`) a little funky. I do like early returns, but returning _with_ the same value each time felt weird. That made me wonder if `finally` could override the return value, and I think the answer is _yes_ (TIL!)

```ts
const write = getAction(...)(args);

try {
  if (someCondition) return;
  if (anotherCondition) return;
}
finally {
  return write;
}
```

not sure if I love this or hate this but it sure is interesting haha

an alternative is a self-invoking function
```ts
const write = getAction(...)(args);

(() => {
  if (someCondition) return;
  if (anotherCondition) return;
})()

return write;
```

anyway, no action needed, just was an interesting thought experiment that taught me something about JS

Reply:
Yeah, not very fond of the multiple early returns either. TIL about `finally` as well haha. Although in this case I prefer to return `write` multiple times just because it makes the intent clearer. I'd love to do it cleaner here but the `calls` type is unknown, so have to do all these checks to make sure it's actually the `userOp` we are expecting.